File: modules/examples/scalacalc/calcservices/src/main/scala/vanadis/modules/examples/scalacalc/calcservices/OperatorWiring.scala
package net.sf.vanadis.modules.examples.scalacalc.calcservices

import scala.collection.mutable
import ext.{Track, Expose}

trait OperatorWiring {

  @Track { val trackedType = classOf[Adder], val remotable = true }
  def adders: mutable.Set[Adder]

  @Track { val trackedType = classOf[Subtractor], val remotable = true }
  def subtractors: mutable.Set[Subtractor]

  @Track { val trackedType = classOf[Multiplier], val remotable = true }
  def multipliers: mutable.Set[Multiplier]

  @Track { val trackedType = classOf[Divisor], val remotable = true }
  def divisors: mutable.Set[Divisor]

  @Expose
  def calculator: PocketCalculator
}

File: modules/examples/javacalc/calcservices/src/main/java/vanadis/modules/examples/javacalc/calcservices/OperatorWiring.java
/*
 * Copyright 2008 Kjetil Valstadsve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package vanadis.modules.examples.javacalc.calcservices;

import vanadis.ext.Expose;
import vanadis.ext.Track;

import java.util.Set;

public interface OperatorWiring {

    @Track(trackedType = Adder.class, remotable = true)
    Set<Adder> getAdders();

    @Track(trackedType = Subtractor.class, remotable = true)
    Set<Subtractor> getSubtractors();

    @Track(trackedType = Multiplier.class, remotable = true)
    Set<Multiplier> getMultipliers();

    @Track(trackedType = Divisor.class, remotable = true)
    Set<Divisor> getDivisors();

    @Expose
    PocketCalculator getPocketCalculator();
}
